File: thinking-orbs-mega/ports/svelte/src/lib/engine/orbits.ts
import { hashD, makeProj, paint, radiusScale, type InkColor } from "./core.js";
import type { Dot, ModeDraw } from "./types.js";

export const drawOrbits: ModeDraw = (
	ctx,
	size,
	t,
	dark,
	o,
	ink: InkColor | undefined,
) => {
	const cx = size / 2;
	const cy = size / 2;
	const radius = (size / 2) * 0.82;
	const project = makeProj(t * 0.12, 0.3, cx, cy, 1);
	const radiusMultiplier = radiusScale(size, o.rsPow ?? 0.6);
	const dots: Dot[] = [];
	const orbitN = o.orbitN ?? 12;
	const ghostN = o.ghostN ?? 40;
	const particles = o.particles ?? 3;

	for (let orb = 0; orb < orbitN; orb++) {
		const h1 = hashD(orb, 1.7);
		const h2 = hashD(orb, 5.2);
		const h3 = hashD(orb, 8.9);
		const orbitRadius = radius * (0.45 + 0.52 * h1);
		const theta = h1 * 2 * Math.PI;
		const phi = Math.acos(2 * h2 - 1);
		const nx = Math.sin(phi) * Math.cos(theta);
		const ny = Math.cos(phi);
		const nz = Math.sin(phi) * Math.sin(theta);
		let ux = -ny;
		let uy = nx;
		const uz = 0;
		const length = Math.max(1e-6, Math.sqrt(ux * ux + uy * uy));
		ux /= length;
		uy /= length;
		const vx = ny * uz - nz * uy;
		const vy = nz * ux - nx * uz;
		const vz = nx * uy - ny * ux;
		const orbitSpeed = (0.25 + 0.55 * h3) * (h3 > 0.5 ? 1 : -1);

		for (let k = 0; k < ghostN; k++) {
			const angle = (k / ghostN) * 2 * Math.PI;
			const [x, y, z] = project(
				(ux * Math.cos(angle) + vx * Math.sin(angle)) * orbitRadius,
				(uy * Math.cos(angle) + vy * Math.sin(angle)) * orbitRadius,
				(uz * Math.cos(angle) + vz * Math.sin(angle)) * orbitRadius,
			);
			const depth = (z / orbitRadius + 1) / 2;
			dots.push({
				x,
				y,
				z,
				r: (o.ghostR ?? 0.9) * radiusMultiplier,
				white: 0.72,
				a: (o.ghostA ?? 0.5) * (0.4 + 0.6 * depth),
			});
		}

		for (let particle = 0; particle < particles; particle++) {
			const angle =
				t * orbitSpeed + (particle / particles) * 2 * Math.PI + h2 * 6;
			const [x, y, z] = project(
				(ux * Math.cos(angle) + vx * Math.sin(angle)) * orbitRadius,
				(uy * Math.cos(angle) + vy * Math.sin(angle)) * orbitRadius,
				(uz * Math.cos(angle) + vz * Math.sin(angle)) * orbitRadius,
			);
			const depth = (z / orbitRadius + 1) / 2;
			dots.push({
				x,
				y,
				z,
				r:
					((o.partR ?? 1.2) + (o.partRDepth ?? 1.6) * depth) * radiusMultiplier,
				white: 0.3 - 0.22 * depth,
			});
		}
	}
	paint(ctx, dots, dark, o.rMin, ink);
};

File: thinking-orbs-mega/ports/gpui/src/engine/orbits.rs
//! Orbits: particles on tilted orbits — the "working" state.

use super::core::{hash_d, make_proj, radius_scale, with_unit_circle, Dot, Frame};
use super::profiles::{count_usize, ModeOpts, MAX_GHOST_N, MAX_ORBIT_N, MAX_PARTICLES};
use std::f32::consts::PI;

pub fn draw_orbits_into(size: f32, t: f32, o: &ModeOpts, out: &mut Frame) {
    let cx = size / 2.0;
    let cy = size / 2.0;
    let r = (size / 2.0) * 0.82;
    let pt = make_proj(t * 0.12, 0.3, cx, cy, 1.0);
    let rs = radius_scale(size, o.rs_pow.unwrap_or(0.6));

    let orbit_n = count_usize(o.orbit_n, 12.0, 0, MAX_ORBIT_N as usize);
    let ghost_n = count_usize(o.ghost_n, 40.0, 0, MAX_GHOST_N as usize);
    let particles = count_usize(o.particles, 3.0, 0, MAX_PARTICLES as usize);

    let dots = &mut out.dots;
    let per_orbit = ghost_n.saturating_add(particles);
    dots.reserve(orbit_n.saturating_mul(per_orbit));
    let ghost_r = o.ghost_r.unwrap_or(0.9) * rs;
    let ghost_a = o.ghost_a.unwrap_or(0.5);
    let part_r = o.part_r.unwrap_or(1.2);
    let part_r_depth = o.part_r_depth.unwrap_or(1.6);

    for orb in 0..orbit_n {
        let orb_f = orb as f32;
        let h1 = hash_d(orb_f, 1.7);
        let h2 = hash_d(orb_f, 5.2);
        let h3 = hash_d(orb_f, 8.9);
        let ro = r * (0.45 + 0.52 * h1);
        let th = h1 * 2.0 * PI;
        let phi = (2.0 * h2 - 1.0).acos();
        // Build a stable plane basis by crossing the normal with whichever
        // world axis is least parallel to it.
        let nx = phi.sin() * th.cos();
        let ny = phi.cos();
        let nz = phi.sin() * th.sin();
        let (mut ux, mut uy, mut uz) = if nz.abs() < 0.9 {
            (-ny, nx, 0.0)
        } else {
            (0.0, -nz, ny)
        };
        let ul = (ux * ux + uy * uy + uz * uz).sqrt().max(1e-6);
        ux /= ul;
        uy /= ul;
        uz /= ul;
        let vx = ny * uz - nz * uy;
        let vy = nz * ux - nx * uz;
        let vz = nx * uy - ny * ux;
        let speed = (0.25 + 0.55 * h3) * if h3 > 0.5 { 1.0 } else { -1.0 };
        let inv_ro = 1.0 / ro;

        // Ghost topology is invariant; reuse its unit-circle samples instead
        // of evaluating 2 × orbit_n × ghost_n trig functions every frame.
        with_unit_circle(ghost_n, |circle| {
            for p in circle {
                let (ca, sa) = (p[0], p[1]);
                let (px, py, z) = pt.project(
                    (ux * ca + vx * sa) * ro,
                    (uy * ca + vy * sa) * ro,
                    (uz * ca + vz * sa) * ro,
                );
                let depth = (z * inv_ro + 1.0) / 2.0;
                dots.push(Dot::new(px, py, z, ghost_r, 0.72).with_a(ghost_a * (0.4 + 0.6 * depth)));
            }
        });
        // the particles doing the work
        for m in 0..particles {
            let a = t * speed + (m as f32 / particles as f32) * 2.0 * PI + h2 * 6.0;
            let (ca, sa) = (a.cos(), a.sin());
            let (px, py, z) = pt.project(
                (ux * ca + vx * sa) * ro,
                (uy * ca + vy * sa) * ro,
                (uz * ca + vz * sa) * ro,
            );
            let depth = (z * inv_ro + 1.0) / 2.0;
            dots.push(Dot::new(
                px,
                py,
                z,
                (part_r + part_r_depth * depth) * rs,
                0.3 - 0.22 * depth,
            ));
        }
    }
}
